### PR TITLE
Add AdminClient.list_groups API

### LINKED ISSUE #223: ListGroups, DescribeGroups support  (for ConsumerOffsetChecker type tool)
Hey I was wondering if there was any way either in the the kafka tools or a script with the confluent kafka python  module to:

- get a list of all consumer groups
- read the current offset of all consumers in a group

When trying to run these commands for ./kafka-consumer-groups.sh it looks like they only work for Java consumers only:

$ ./kafka-consumer-groups.sh  --new-consumer --bootstrap-server  XXXX:6667  --describe --group 729a8c43-fd73-4fe6-b8e0-b0d73096f30f
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

./kafka-consumer-groups.sh  --new-consumer --bootstrap-server  xxxx:6667  --list
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).
logstash
...

Thanks,

Shan

### PR DESCRIPTION
**Pull request by messense**
Closes #223

Not sure how to add test cases for this feature.

### CODE REVIEW

**Comment by messense** [comment]
[clabot:check]

File: confluent_kafka/src/Metadata.c [hunk 1]
```diff
@@ -190,6 +190,33 @@ c_topics_to_py (Handle *self, const rd_kafka_metadata_topic_t *c_topics,
...
+        PyObject *key;
+
+        broker = PyObject_CallObject(BrokerMetadata_type, NULL);
+        if (!broker)
+                return NULL;
+
+        key = cfl_PyInt_FromInt(c_broker.id);
+
+        if (PyObject_SetAttrString(broker, "id", key) == -1) {
+                Py_DECREF(key);
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:205]: broker is leaked here

File: confluent_kafka/src/Metadata.c [hunk 2]
```diff
@@ -190,6 +190,33 @@ c_topics_to_py (Handle *self, const rd_kafka_metadata_topic_t *c_topics,
...
+                return NULL;
+        }
+        Py_DECREF(key);
+
+        if (cfl_PyObject_SetString(broker, "host",
+                                        c_broker.host) == -1)
+                return NULL;
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:212]: broker is leaked

File: confluent_kafka/src/Metadata.c [hunk 3]
```diff
@@ -190,6 +190,33 @@ c_topics_to_py (Handle *self, const rd_kafka_metadata_topic_t *c_topics,
...
+        if (cfl_PyObject_SetInt(broker, "port",
+                                (int)c_broker.port) == -1)
+                return NULL;
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:215]: dito

File: confluent_kafka/src/Metadata.c [hunk 4]
```diff
@@ -405,3 +419,215 @@ const char list_topics_doc[] = PyDoc_STR(
         " :param float timeout: Maximum response time before timing out, or -1 for infinite timeout.\n"
         " :rtype: ClusterMetadata \n"
         " :raises: KafkaException \n");
+
+
+static PyObject *
+c_group_members_to_py(Handle *self, const struct rd_kafka_group_member_info *c_members,
+                      int member_cnt)
+{
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:427]: style nit: open-brace should go on the function params line, like so:
`... int member_cnt) {`

File: confluent_kafka/src/Metadata.c [hunk 5]
```diff
@@ -405,3 +419,215 @@ const char list_topics_doc[] = PyDoc_STR(
...
+        if (!groups)
+                goto err;
+        for (i = 0; i < group_list->group_cnt; i++) {
+                PyObject *group, *error, *broker, *members;
+
+                group = PyObject_CallObject(GroupMetadata_type, NULL);
+                if (!group)
+                        goto err;
+
+                if (cfl_PyObject_SetString(group, "group",
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:524]: Maybe this should be called "id" or "group_id" instead.

File: confluent_kafka/src/Metadata.c [hunk 6]
```diff
@@ -405,3 +419,215 @@ const char list_topics_doc[] = PyDoc_STR(
...
+        if (group_list != NULL) {
+                rd_kafka_group_list_destroy(group_list);
+        }
+        return result;
+}
+
+const char list_groups_doc[] = PyDoc_STR(
+        ".. py:function:: list_groups([group=None], [timeout=-1])\n"
+        "\n"
+        " Request Metadata from cluster.\n"
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:626]: "Request Group Metadata from cluster."

File: confluent_kafka/src/Metadata.c [hunk 7]
```diff
@@ -405,3 +419,215 @@ const char list_topics_doc[] = PyDoc_STR(
...
+        " This method provides the same information as"
+        " listGroups(), describeGroups() in the Java Admin client.\n"
+        "\n"
+        " :param str group: If specified, only reqeust info about this group, else return for all groups in cluster"
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:630]: typo: request
missing period at end of line

File: confluent_kafka/src/Metadata.c [hunk 8]
```diff
@@ -405,3 +419,215 @@ const char list_topics_doc[] = PyDoc_STR(
...
+        " :param float timeout: Maximum response time before timing out, or -1 for infinite timeout.\n"
+        " :rtype: GroupMetadata \n"
```
Comment (edenhill) [inline confluent_kafka/src/Metadata.c:632]: Remove space before newline, here and on the line below

File: examples/adminapi.py [hunk 9]
```diff
@@ -274,6 +274,19 @@ def example_list(a, args):
                   " isrs: {} errstr: {}".format(p.id, p.leader, p.replicas,
                                                 p.isrs, errstr))
 
+    groups = a.list_groups(timeout=10)
```
Comment (edenhill) [inline examples/adminapi.py:277]: Add two specific command options and dont include them in this metadata listing.
 - ` .. list groups`
 - ` .. list group <group-id>`

File: tests/test_Admin.py [hunk 10]
```diff
@@ -36,6 +36,11 @@ def test_basic_api():
     except KafkaException as e:
         assert e.args[0].code() in (KafkaError._TIMED_OUT, KafkaError._TRANSPORT)
 
+    try:
+        a.list_groups(timeout=0.2)
+    except KafkaException as e:
+        assert e.args[0].code() in (KafkaError._TIMED_OUT, KafkaError._TRANSPORT)
+
```
Comment (edenhill) [inline tests/test_Admin.py:43]: Add an integration test (tests/integration/..) too where you:
 - produce some messages to a topic
 - start two consumers
 - wait for them to consume and commit
 - then call list_groups() both with and without the group name
 - verify group has two members

**Review by edenhill — CHANGES_REQUESTED** [review]
Great stuff!

File: tests/test_Admin.py (hunk 10 above)
Comment (messense) [reply tests/test_Admin.py:43]: Unfortunately I can't get the integration tests running on my mac, it errors with

```
unable to resolve docker endpoint: context "/Users/xxx/Projects/confluent-kafka-python/tests/docker/docker-compose.yaml" does not exist
```

But the file does exists

Comment (edenhill) [reply tests/test_Admin.py:43]: I think you can just run the integration_test.py script directly by creating a testconf.json file with bootstrap.servers set

Comment (messense) [reply tests/test_Admin.py:43]: It works.

```
python tests/integration/integration_test.py --admin tests/integration/testconf.json
Using confluent_kafka module version 1.5.0 (0x1050000)
Using librdkafka version 1.5.0 (0x10500ff)
Testing: ['admin']
Brokers: localhost:9092
Topic prefix: test-topic
============================== Verifying Admin API ==============================
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [0] at offset 1
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [1] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [3] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [4] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [2] at offset 1
EOF reached for all assigned partitions: exiting
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [0] at offset 1
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [1] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [3] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [4] at offset 0
Reached end of test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c [2] at offset 1
EOF reached for all assigned partitions: exiting
Topic test-topic_admin_c63da45f-4f28-4160-9d28-8757ec70147c marked for deletion
============================== Done ==============================
```

File: confluent_kafka/admin/__init__.py [hunk 11]
```diff
@@ -573,3 +573,57 @@ def __repr__(self):
...
+    For more information on metadata format, see
+    https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-GroupMembershipAPI
+
+    This class is typically not user instantiated.
+
+    :ivar str id: Member id (generated by broker)
+    :ivar str client_id: Client id
+    :ivar str client_host: Client hostname
+    :ivar bytes metadata: Member metadata(binary), format depends on protocol type
+    :ivar bytes assignment: Member assignment(binary), format depends on protocol type
```
Comment (messense) [inline confluent_kafka/admin/__init__.py:590]: Would be great to have APIs to parse `metadata` and `assignment`.

cc https://github.com/edenhill/librdkafka/issues/2643

File: tests/integration/integration_test.py [hunk 12]
```diff
@@ -1110,6 +1110,61 @@ def verify_admin():
...
+    consume_messages(group2)
+    # list_groups without group argument
+    groups = set(group.id for group in a.list_groups(timeout=10))
+    assert group1 in groups, "Consumer group {} not found".format(group1)
+    assert group2 in groups, "Consumer group {} not found".format(group2)
+    # list_groups with group argument
+    groups = set(group.id for group in a.list_groups(group1))
+    assert group1 in groups, "Consumer group {} not found".format(group1)
+    groups = set(group.id for group in a.list_groups(group2))
+    assert group2 in groups, "Consumer group {} not found".format(group2)
```
Comment (edenhill) [inline tests/integration/integration_test.py:1166]: :clap:

File: confluent_kafka/admin/__init__.py (hunk 11 above)
Comment (edenhill) [reply confluent_kafka/admin/__init__.py:590]: That'd be great, but could be provided as a separate python utility package, perhaps auto-generated from the AK json protocol specs.

File: setup.py [hunk 13]
```diff
@@ -25,7 +25,8 @@
 
 SCHEMA_REGISTRY_REQUIRES = ['requests']
 
-AVRO_REQUIRES = ['fastavro>=0.23.0',
+AVRO_REQUIRES = ['fastavro>=0.23.0,<1.0;python_version<"3.0"',
```
Comment (edenhill) [inline setup.py:28]: Thanks for fixing this.

**Review by edenhill — CHANGES_REQUESTED** [review]
Solid stuff!
I think it would be useful to print the protocol and type in the example.
After that we're good to go

File: examples/adminapi.py [hunk 14]
```diff
@@ -252,27 +252,39 @@ def example_list(a, args):
...
+        for g in groups:
+            if g.error is not None:
+                errstr = ": {}".format(t.error)
             else:
                 errstr = ""
 
-            print("partition {} leader: {}, replicas: {},"
-                  " isrs: {} errstr: {}".format(p.id, p.leader, p.replicas,
-                                                p.isrs, errstr))
+            print(" \"{}\" with {} member(s){}".format(g, len(g.members), errstr))
```
[the remark this answers is not in the document]
Comment (messense) [reply examples/adminapi.py]: Added in https://github.com/confluentinc/confluent-kafka-python/pull/948/commits/c2effef3f8eca28c5cd318828580961f2522f5dc.

**Review by edenhill — APPROVED** [review]

**Comment by edenhill** [comment]
Great work on this, thank you!

**Comment by RoyAtanu** [comment]
@edenhill Is there a timeline when this change will make to a release?

**Comment by edenhill** [comment]
We'll release v1.6.0 after the holidays.

File: confluent_kafka/admin/__init__.py (hunk 11 above)
Comment (snagafritz) [reply confluent_kafka/admin/__init__.py:590]: I'm interested in decoding this data... is this work in progress somewhere? Or maybe point me to the spec you're referring to (I couldn't find it with limited googling) and I can take a crack at it.

Comment (snagafritz) [reply confluent_kafka/admin/__init__.py:590]: Found the spec:
https://kafka.apache.org/protocol

Comment (JjudeGotLemon) [reply confluent_kafka/admin/__init__.py:590]: @snagafritz  What's you solution to decode it? Could you share it out?

Comment (messense) [reply confluent_kafka/admin/__init__.py:590]:
```python
def parse_member_assignment(buf):
    '''Parse Kafka group assignment'''
    import struct
    assigns = []
    if not buf:
        return assigns
    index = struct.calcsize('>hi')
    _version, assign_len = struct.unpack('>hi', buf[0:index])
    for _ in range(assign_len):
        size = struct.calcsize('>h')
        str_len = struct.unpack('>h', buf[index:index + size])[0]
        index += size
        topic = buf[index: index + str_len].decode('utf-8')
        index += str_len;
        size = struct.calcsize('>i')
        part_len = struct.unpack('>i', buf[index:index + size])[0]
        index += size
        parts = []
        for _ in range(part_len):
            part = struct.unpack('>i', buf[index:index + size])[0]
            parts.append(part)
            index += size
        assigns.append((topic, parts))
    return assigns
```

Comment (JjudeGotLemon) [reply confluent_kafka/admin/__init__.py:590]: @messense Very helpful! Thanks

**Pull request merged**